near-plugins-derive/src/upgradable.rs

```diff
@@ -21,19 +22,29 @@ pub fn derive_upgradable(input: TokenStream) -> TokenStream {
         .code_storage_key
         .unwrap_or_else(|| "__CODE__".to_string());
 
+    let allowed_timestamp_storage_key = opts
+        .allowed_timestamp_storage_key
+        .unwrap_or_else(|| "__ALLOWED_TIMESTAMP__".to_string());
+
     let output = quote! {
         #[near_bindgen]
         impl Upgradable for #ident {
-            fn up_storage_key(&self) -> Vec<u8>{
+            fn up_storage_key(&self) -> Vec<u8> {
                 (#code_storage_key).as_bytes().to_vec()
             }
 
+            fn up_allowed_timestamp_storage_key(&self) -> Vec<u8> {
+                (#allowed_timestamp_storage_key).as_bytes().to_vec()
+            }
+
             #[#cratename::only(owner)]
-            fn up_stage_code(&mut self, #[serializer(borsh)] code: Vec<u8>) {
+            fn up_stage_code(&mut self, #[serializer(borsh)] code: Vec<u8>,  #[serializer(borsh)] delay_timestamp: u64) {
```

Comment:
The staging time should be independent from the submission of the code. Otherwise, this protection is meaningless. Changing staging time should require higher privileges than staging code. This way users can trust that code won't change without having a notice period.

See a similar proposal with this feature here: https://github.com/near/NEPs/pull/123